feat: Add RT apps snap packaging to the existing tutorial (#99)
canonical/real-time-ubuntu-docs: docs/tutorial/7-snaping.md

# Snapping your real-time app

Now that we have our real-time app working, it's time to package it.
This section describes how to package your real-time app as a [snap package][snapcraft.io].

## Installing requirements

Some dependencies are required when building snap packages:

- Confirm that you have [snapd][snapd] installed by running snap version:
  If you're using Ubuntu, it should come pre-installed. 
  Otherwise, refer to the [snapd installation docs][install_snapd].

- Install the [snapcraft snap][snapcraft_docs]:
  ```shell
    sudo snap install --classic snapcraft
  ```

## Packaging

The first step in creating a snap is to generate the default template by running `snapcraft init`:

```console
$ snapcraft init
Go to https://docs.snapcraft.io/the-snapcraft-format/8337 for more information about the snapcraft.yaml format.
Successfully initialised project.
```

> This command creates a snap directory containing the snapcraft.yaml file.

Reviewer: ```suggestion
This command creates a `snap` directory containing the `snapcraft.yaml` file.
```
For consistency, as we codify the file name later too.

version: '0.1' # just for humans, typically '1.2+git' or '1.3.2'
summary: Single-line elevator pitch for your amazing snap # 79 char long summary
description: |
  This is my-snap's description. You have a paragraph or two to tell the
  most important story about your snap. Keep it under 100 words though,
  we live in tweetspace and your description wants to look good in the snap
  store.

grade: devel # must be 'stable' to release into candidate/stable channels
confinement: devmode # use 'strict' once you have the right plugs and slots

parts:
  my-part:
    # See 'snapcraft plugins'
    plugin: nil
```
You may also want to review the [snapcraft.yaml schema][snapcraft_yaml].
Which serves as a reference for understanding what is possible in the `snapcraft.yaml` file.

Your working directory should now look like this:

```console
$ tree
.
├── cfs.c
├── edf.c
├── fifo.c
├── lock.c
├── snap
│   └── snapcraft.yaml
└── thread-affinity.c

2 directories, 6 files
```

To streamline the build process for our real-time apps, we should create a Makefile.
This will also help with packaging.

Let's create one:

```{literalinclude} Makefile
:language: make

> ```

Reviewer 2: Is the make file really necessary?

Author: Actually, no, but as I mention in the tutorial, using a Makefile makes life much easier by reducing the parts definition to just three lines:

```yaml
parts: 
  src: 
    plugin: make 
    source: . 
```

Without it, I’d need to spend much more effort explaining the fine details of overriding the build process, staging the right packages, and handling dependencies manually.

This approach also gives us the opportunity to introduce users to Snapcraft plugins and demonstrate how simple it can be to use one to package an application.

Since almost every `C` project includes a Makefile, most developers familiar with `C`  will already be comfortable with it. By doing this, we shift the complexity away from the packaging side and into something the user already knows.

This is also exactly what **I would do** if I were developing a `C` application to be snapped.

So, unless there's a strong reason to avoid it, I’d prefer to keep the Makefile. 
If you see any major concerns, I’d appreciate your clarification.

Reviewer 2: > but as I mention in the tutorial, using a Makefile makes life much easier by **reducing the parts definition to just three lines**

You did not mention this in the tutorial. 

But yes, I understand that it is good to keep.

> We'll also make some modifications to the YAML manifest file:

Reviewer: This is the first time we call this the "manifest file". Maybe rather stick to the same naming throughout, and not introduce new terms. Or call it the manifest file from the start.

```suggestion
We'll also make some modifications to the `snapcraft.yaml` file:
```

Author: It doesn't introduce new terms. The term "manifest file", or just "manifest" it's broadly used across the community. I've seeing this many times in the Matrix chat of Snapcraft.

So the user needs to be familiarized with the both ways.

Reviewer 2: I also think it should be `snapcraft.yaml`.

It is true that developers call it the manifest file, but there is no snap documentation for anything called a YAML manifest. The reference uses the file name: https://snapcraft.io/docs/snapcraft-yaml-schema

Launching managed ubuntu 24.04 instance...
Creating new instance from remote
Creating new base instance from remote
Creating new instance from base instance
Starting instance
Starting snapcraft, version 8.7.3
Logging execution to '/tmp/snapcraft.log'
Initialising lifecycle
Installing build-packages
Installing build-snaps
Pulling src
Building src
:: + make -j12
:: gcc -Wall -Wextra -Wno-unused-result -O2 -o cfs cfs.c
:: gcc -Wall -Wextra -Wno-unused-result -O2 -o edf edf.c
:: gcc -Wall -Wextra -Wno-unused-result -O2 -o fifo fifo.c
:: gcc -Wall -Wextra -Wno-unused-result -O2 -o lock lock.c
:: gcc -Wall -Wextra -Wno-unused-result -O2 -o thread-affinity thread-affinity.c
:: + make -j12 install DESTDIR=/root/parts/src/install
:: mkdir -p /root/parts/src/install/bin
:: install -m 755 cfs edf fifo lock thread-affinity /root/parts/src/install/bin
Staging src
Priming src
Packing...
Reading snap metadata...
Running linters...
Running linter: classic
Running linter: library
Creating snap package...
Packed rt-app_0.1_amd64.snap
```

The first time you build a snap, it may take longer because Snapcraft also downloads and installs a [build provider][build_provider] (as seen in the logs).

> It also provisions a [LXD container][lxd_container].

Reviewer: ```suggestion
It also provisions an [LXD container][lxd_container].
```

squashfs-root/
├── bin
│   ├── cfs
│   ├── edf
│   ├── fifo
│   ├── lock
│   └── thread-affinity
└── meta
    ├── gui
    └── snap.yaml

4 directories, 6 files
```

Now that we've confirmed that the real-time app binaries are present, it's time to define the apps.

### Defining snap apps

A Snap doesn't necessarily need to have apps defined—it can provide only files such as libraries or binaries.
These are generally referred to as content Snaps, as they use the [content interface][content_interface].
That's why it's possible to build a Snap successfully without defining any apps.

The simplest app definition includes the app name (which is the name of the object in the apps list) and the `command` property.
Since the apps are inside the `bin` directory, we need to define the relative path to them:

```yaml
apps:
  cfs:
    command: bin/cfs

  edf:
    command: bin/edf

  fifo:
    command: bin/fifo

  lock:
    command: bin/lock

  thread-affinity:
    command: bin/thread-affinity

```

After defining the apps, we can rebuild the Snap by running `snapcraft -v` again.
The second build is faster than the first one because the necessary dependencies are already available.

> As well the necessary resources already created by the build provider (which is a lxd container in this case).

Reviewer: ```suggestion
Also the necessary resources are already created by the build provider (which is a lxd container in this case).
```

We can then install the Snap using:

```console
$ sudo snap install rt-app_0.1_amd64.snap
error: cannot find signatures with metadata for snap "rt-app_0.1_amd64.snap"
```

The command fails because we're installing a locally built artifact.

> According to the [Snap install modes document][snap_install_modes], when developing and testing a Snap, it's recommended to install it in `devmode`:

Reviewer: Is it `--devmode` that we require or `--dangerous`?

Author: From [the documentation](https://snapcraft.io/docs/install-modes):

> --classic, for classic confinement, which provides the same system access as traditional types of packages.
> --dangerous, for testing unsigned, unvetted, and risky snaps.
> --devmode, for development testing and viewing the log output.
> --jailmode, for forcing strict confinement.

So, if the user is developing the application, it's recommended to install in `--devmode`. You wouldn't be able to install with `--dangerous` because the `confinement` field is set as `devmode` at this point.

While the first approach is recommended for completeness and accuracy, the second approach is useful when initially confining the application.

The complexity of the application determines the best approach.
If the application has multiple configurations, each triggering different execution paths, a full source code analysis is preferable.
However, if the application has a single, fixed execution path, debugging tools alone may be sufficient.

A variety of resources and tools exist for [debugging Snaps][debug_snaps].
The recommended tool for debugging and initial confinement is [snappy-debug].

To install it:

```console
$ sudo snap install snappy-debug
snappy-debug 0.36-snapd2.59.4 from Canonical✓ installed
```

Now, we need two terminals: one to run `snappy-debug` and another to execute the `rt-app` Snap applications.
Running `snappy-debug` presents the following message:

```console
$ snappy-debug
INFO: Following '/var/log/syslog'. If have dropped messages, use:
INFO: $ sudo journalctl --output=short --follow --all | sudo snappy-debug
```

To ensure no log messages are lost, it is recommended to use the suggested `journalctl` command.

> As well running with `sudo` since accessing system logs requires privileged access:

Reviewer: ```suggestion
It needs to be run with `sudo` since accessing system logs requires privileged access:
```

process-control  rt-app:process-control  :process-control  manual
```

#### Analyzing `fifo` and `lock` applications

Repeat the same process done for `cfs` and `edf` applications on this one. 
Like occurred for `cfs`, when running `fifo` application no reports were generated by `snappy-conf`.
The same happens for the `lock` application, no reports were generated.

#### Analyzing `thread-affinity` application

Repeating the process for `thread-affinity` something different happens on the logs:

- Application output:
```console
$ rt-app.thread-affinity
thread1 priority: 0
thread2 priority: 0
Calls made on thread1: 100000
Calls made on thread2: 1
```

- Debug output:
```
= Seccomp =
Time: Mar 27 21:04:22
Log: auid=1000 uid=1000 gid=1000 ses=119 subj=snap.rt-app.thread-affinity pid=40684 comm="thread-affinity" exe="/snap/rt-app/x2/bin/thread-affinity" sig=0 arch=c000003e 203(sched_setaffinity) compat=0 ip=0x733af5a4d6e1 code=0x7ffc0000
Syscall: sched_setaffinity
Suggestion:
* ignore the denial if the program otherwise works correctly (unconditional sched_setaffinity is often just noise)
```

We see a seccomp report, but no suggested interface to use.

> Instead a message saying to just ignore the denial report.

Reviewer: ```suggestion
Instead there is a message suggesting to just ignore the denial report.
```

> This happens because there are some denials coming the security policies defined in the snap sandboxing may not affect the functionality of an application.

Reviewer: ```suggestion
This happens because there are some denials coming from the security policies defined in the snap sandboxing, which potentially does not affect the functionality of an application.
```

This is the case for this app.

### Concluding the strict confinement

Now that the required interfaces are defined, change the `confinement` property to `strict`:

```yaml
confinement: strict
```

Then:
  - Build the snap like done before with `snapcraft`.

>   - Install it. This time, since we aren't using `devmode` anymore, it's possible to install it in a confined way.
>     But because the snap isn't being installed from a snap store, there are no assertions files.
>     This means that it's necessary to acknowledge the **dangerous** nature of it.

Reviewer: ```suggestion
  - Install it. This time, since we aren't using `devmode` anymore, it's possible to install it in a confined way.
    Because the snap isn't being installed from a snap store, there is no way to fully trust it.
    It is therefore necessary to acknowledge the **dangerous** nature of it.
```

    So we run our install command with the `--dangerous` flag.

    ```console
    $ sudo snap install ./rt-app_0.1_amd64.snap --dangerous
    rt-app 0.1 installed
    ```

  - Connect the `process-control` interface as shown before.

>     It may be already connected if during the iterations the snap wasn't uninstalled.

Reviewer: ```suggestion
    It might still be connected if the snap wasn't uninstalled during the development steps.
```

  - Run all the applications.
    You may see some logs on the `snappy-debug` terminal.

>     It's necessary to pain attention on the field `profile=` for [AppArmor][apparmor] denials and `exe="` for Seccomp denials. 

Reviewer: ```suggestion
    It's necessary to pay attention to the field `profile=` for [AppArmor][apparmor] denials and `exe="` for Seccomp denials. 
```

    If it says: `profile="/usr/lib/snapd/snap-confine"`, it's related to the `snap-confine` component and not about the `rt-app` snap.

>     To know more about it, consider take a look on this reference about [the snap system architecture][snap_sys_arch].

Reviewer: ```suggestion
    To learn more about this, take a look at this reference of [the snap system architecture][snap_sys_arch].
```

> Then you've finished the development of your snap, you can now change the `grade` field to `stable`, so it can be published on the store.

Reviewer: ```suggestion
You have now finished the development of your snap.
You can now change the `grade` field to `stable`, so it can be published on the Snap Store.
```

### Conclusion

The final `snapcraft.yaml` manifest file should look like this:

```{literalinclude} snapcraft.yaml
:language: yaml
```

> The next steps that you may consider to do:

Reviewer: ```suggestion
The following How-To guides are recommended as next steps:
```

Author: Actually, the first link _might_ be a `How-To` (I'm not sure because it doesn't show the position of the link into the document tree on the left, but It really seems to be a how-to).

But the second one, it's actually an explanation. There are many documents about brand store, not sure which one is more indicated in this situation.
